libflux/Polymorphic_Labels.md

```diff
@@ -0,0 +1,103 @@
+The types that fit into a record field implement the new `Label` kind to ensure that unexpected types like `Int` are an error when used in a record field.
+
+To make the change as transparent as possible string literals now get the new `Label("literal")` type instead of `String` making `"abc": Literal("abc")`. While this means that two different string literals will have different types we must still treat `Label` types as strings in most cases. Consider
+
+```flux
+if b then "a" else "b"
+```
+
+The branches will have the `Label("a")` and `Label("b")` respectively, which we still need to unify successfully and the resulting type of the expression should still be a `string`. As long as we limit  the uses of label types to direct uses in calls (like in `fill` above). We may keep the same type checking behaviour by treating `Label` as a `String` types in every instance except when type checking function calls, in which case we specialize unification to a subsumption check where we allow some limited sub-typing.
+
+```flux
+// The original fill call is allowed such that `'column` is inferred to be `Label("a")`
+fill(column: "a", value: 0)
+
+// We should (probably) still allow dynamic strings for backwards compatibility sake so the `string` type
+// also implements the `Label` Kind
+c = "a" + ""
+fill(column: c, value: 0)
+
+builtin add : (x: A, y: A) => A where A: Addable
+
+// This must still be allowed, however a naive implementation would first infer `A <=> Label("a")`
+// and then fail to unify `Label("a") <=> Label("b")`. If we keep treating `Label(..)` as a string except in cases where it unifies to a variable that actually has the `Label` type.
+add(x: "a", y: "b")
+
+builtin func : (opt: string) => int
+
+// This will work since `Label("a")` is a sub type of `string`
+func(opt: "a")
+
+// Possible extension where we only allow some specific labels to be passed in
+builtin func : (opt: "option1" | "option2") => int
+
+// "option1" is an allowed option
+func(opt: "option1")
+// However "option3" is not allowed
+func(opt: "option3")
+// However passing in a dynamic string could be disallowed (`string` is to general)
+o = "option" + ""
+func(opt: o)
+```
+
+Row polymorphism introduces another wrinkle to the implementation. When fields aren't necessarily know at the point that `unify` are called we aren't able to match the field of either side.
+
+```flux
+// Unifying know fields works regardless of the order that they are defined in
+{ a: int, b: string } <=> { b: string, a: int }
+
+// Should 'column unify against `a` or `b` field on the left side?
+{ 'column: A } <=> { b: string, a: int }
+```
+
+There may be a consistent way to unify these records in the face of type variables, however an easy workaround would be to delay the unification of records with unknown fields until they have been resolved, at which point they can unify normally. If there is a field that is still unknown when type checking is done we can designate that as an error.
```

Comment:
To me delaying unification until records with unknown fields have been resolved makes the most sense. Its how I reason about this logic in the first place. Are there any drawbacks to delaying? Multiple passes over the semantic graph?

Reply:
It adds some complexity, and could possibly give worse error messages (though as long as we take care to store all the context we need for the errors, it shouldn't be an issue).

There may be some performance implications as well as we need to know/decide when to evaluate these delayed constraints. Constraints could (in theory at least) depend on each other in complex ways which we would need to untangle to resolve them all

A bit of a tangent, but Rust's typechecker has an largely equivalent system of "obligations" which it brute forces when solving them, I tried to improve that with https://github.com/rust-lang/rust/pull/69218 which would make them know more precisely when they would be solveable, but the overhead was to much).